### ISSUE TITLE
Run NOS3 Example with COSMOS and 42 on different VM

### ISSUE THREAD

**Issue by kabarr** [nasa/nos3#83]
I've created 2 VMs and successfully changed settings to run COSMOS and 42 on one machine and the Sims on another.  I've run into a snag that may not be possible to fix.  The truth42Sim data is all 0's in the sim, and the 42 satellite location is "static".  I suspect this is due to me not understanding all the connections between 42 and the sim.  Is it possible to run 42 on a different machine?

**Comment by jlucas9** [comment]
Hi @kabarr,
It should certainly be possible to run them in different VMs. I'm working now on dockerizing the simulators in issue #78, but it's a slow process. 42 should be directly feeding data to the simulators via the [sim_common/src/sim_data_42_socket_provider.cpp](https://github.com/nasa-itc/sim_common/blob/e3eb4d7fb3347e838a91055722548ac4c86cc461/src/sim_data_42socket_provider.cpp) interface. The simulators will timeout if they cannot connect to 42 if I remember correctly - maybe that's what's occurring?

**Comment by kabarr** [comment]
It appears that the 42 truth data is all 0's.  I may not be running it on the correct VM.

VM1: With Flight SW and Sims:
***@***.***

VM2 with 42 Running (sat position does not update:
***@***.***

From: John P. Lucas ***@***.***>
Sent: Friday, September 23, 2022 8:15 AM
To: nasa/nos3 ***@***.***>
Cc: Barr, Kathryn A (MSFC-HP25)[MO&I] ***@***.***>; Mention ***@***.***>
Subject: [EXTERNAL] Re: [nasa/nos3] Run NOS3 Example with COSMOS and 42 on different VM (Issue #83)


Hi @kabarr<https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fkabarr&data=05%7C01%7Ckathryn.a.barr%40nasa.gov%7Cb7261bb2e3684a91aa9a08da9d65a514%7C7005d45845be48ae8140d43da96dd17b%7C0%7C0%7C637995357133077466%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=VCv3XiAV%2B8GDR78zXmv4%2FGEGCLlGGB8zWVpi80%2B55lA%3D&reserved=0>,
It should certainly be possible to run them in different VMs. I'm working now on dockerizing the simulators in issue #78<https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnasa%2Fnos3%2Fissues%2F78&data=05%7C01%7Ckathryn.a.barr%40nasa.gov%7Cb7261bb2e3684a91aa9a08da9d65a514%7C7005d45845be48ae8140d43da96dd17b%7C0%7C0%7C637995357133077466%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=S8ZuXxdgwgiBp85TSS1hOGimZtX1ru0COYnD4BXy9PY%3D&reserved=0>, but it's a slow process. 42 should be directly feeding data to the simulators via the sim_common/src/sim_data_42_socket_provider.cpp<https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnasa-itc%2Fsim_common%2Fblob%2Fe3eb4d7fb3347e838a91055722548ac4c86cc461%2Fsrc%2Fsim_data_42socket_provider.cpp&data=05%7C01%7Ckathryn.a.barr%40nasa.gov%7Cb7261bb2e3684a91aa9a08da9d65a514%7C7005d45845be48ae8140d43da96dd17b%7C0%7C0%7C637995357133077466%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=yGExFSG%2Fa9X2f%2FntvJa%2FWgWGjGXVMeP%2FScRGi9%2BhlHQ%3D&reserved=0> interface. The simulators will timeout if they cannot connect to 42 if I remember correctly - maybe that's what's occurring?

-

**Comment by jlucas9** [comment]
Quick update - I have been able to duplicate this issue on my side and am working with some other experts to debug!

**Comment by jlucas9** [comment]
^ From the `nos3#83` branch 7a66dab:
* `vagrant up` to deploy the VM
* `vagrant halt` to shut it down
* Inside the Oracle VM VirtualBox Manager
  * Tools > Network > Add
    * Network Name: NOS3
    * Network CIDR: 192.168.100.0/24
  * Edit the `nos3_1.5.2` settings
    * Network > Adapter 1 > NAT Network > NOS3
  * Manually cloned the `nos3_1.5.2` VM in the Oracle VM VirtualBox Manager
  * Updated cloned VM settings:
    * Network > Adapter 1 > Advanced > Adapter Type > Intel PRO/1000 MT Server
    * Note that this is different than the default provided to ensure it gets recognized as a different on the network

To summarize I've got:
* NOS3 NAT setup in VirtualBox
* VM0 - FSW and Simulators
* VM1 - 42 and COSMOS (not setup)

I am getting valid 42 data in the truth simulator. If we want to get COSMOS running too in this branch let me know and I can fix it up.

**Comment by kabarr** [comment]
Thanks John!  I'll give this a try - I was out last week (forgot to set my "away notice...).

Best Regards,
Kathy

From: John P. Lucas ***@***.***>
Sent: Tuesday, October 11, 2022 11:52 AM
To: nasa/nos3 ***@***.***>
Cc: Barr, Kathryn A (MSFC-HP25)[MO&I] ***@***.***>; Mention ***@***.***>
Subject: [EXTERNAL] Re: [nasa/nos3] Run NOS3 Example with COSMOS and 42 on different VM (Issue #83)


^ From the nos3#83 branch 7a66dab<https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnasa%2Fnos3%2Fcommit%2F7a66dab0ae9150b06324e1631013c0f06c4c7998&data=05%7C01%7Ckathryn.a.barr%40nasa.gov%7C183afbe84865438e04f208daaba8de45%7C7005d45845be48ae8140d43da96dd17b%7C0%7C0%7C638011039023810596%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=pcHNt1dLINCsPgspnhNXy071%2FcmRE90xQiiVFtnPglc%3D&reserved=0>:

  *   vagrant up to deploy the VM
  *   vagrant halt to shut it down
  *   Inside the Oracle VM VirtualBox Manager
     *   Tools > Network > Add
        *   Network Name: NOS3
        *   Network CIDR: 192.168.100.0/24
     *   Edit the nos3_1.5.2 settings
        *   Network > Adapter 1 > NAT Network > NOS3
     *   Manually cloned the nos3_1.5.2 VM in the Oracle VM VirtualBox Manager
     *   Updated cloned VM settings:
        *   Network > Adapter 1 > Advanced > Adapter Type > Intel PRO/1000 MT Server
        *   Note that this is different than the default provided to ensure it gets recognized as a different on the network

To summarize I've got:

  *   NOS3 NAT setup in VirtualBox
  *   VM0 - FSW and Simulators
  *   VM1 - 42 and COSMOS (not setup)

I am getting valid 42 data in the truth simulator. If we want to get COSMOS running too in this branch let me know and I can fix it up.

-

**Comment by kabarr** [comment]
I was able to review the changes in the #83 branch and update the 2 VMs I had already created.  I'm now seeing Truth data on the machine with FSW/Sims and the 42 is updating on the VM with COSMOS.  Thanks!

From: John P. Lucas ***@***.***>
Sent: Tuesday, October 11, 2022 11:52 AM
To: nasa/nos3 ***@***.***>
Cc: Barr, Kathryn A (MSFC-HP25)[MO&I] ***@***.***>; Mention ***@***.***>
Subject: [EXTERNAL] Re: [nasa/nos3] Run NOS3 Example with COSMOS and 42 on different VM (Issue #83)


^ From the nos3#83 branch 7a66dab<https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnasa%2Fnos3%2Fcommit%2F7a66dab0ae9150b06324e1631013c0f06c4c7998&data=05%7C01%7Ckathryn.a.barr%40nasa.gov%7C183afbe84865438e04f208daaba8de45%7C7005d45845be48ae8140d43da96dd17b%7C0%7C0%7C638011039023810596%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=pcHNt1dLINCsPgspnhNXy071%2FcmRE90xQiiVFtnPglc%3D&reserved=0>:

  *   vagrant up to deploy the VM
  *   vagrant halt to shut it down
  *   Inside the Oracle VM VirtualBox Manager
     *   Tools > Network > Add
        *   Network Name: NOS3
        *   Network CIDR: 192.168.100.0/24
     *   Edit the nos3_1.5.2 settings
        *   Network > Adapter 1 > NAT Network > NOS3
     *   Manually cloned the nos3_1.5.2 VM in the Oracle VM VirtualBox Manager
     *   Updated cloned VM settings:
        *   Network > Adapter 1 > Advanced > Adapter Type > Intel PRO/1000 MT Server
        *   Note that this is different than the default provided to ensure it gets recognized as a different on the network

To summarize I've got:

  *   NOS3 NAT setup in VirtualBox
  *   VM0 - FSW and Simulators
  *   VM1 - 42 and COSMOS (not setup)

I am getting valid 42 data in the truth simulator. If we want to get COSMOS running too in this branch let me know and I can fix it up.

-

**Comment by jlucas9** [comment]
Final step for this issue is to (from the new `dev` branch 59e8ec6) recreate the above and get COSMOS command and telemetry working. @msuder

**Comment by kabarr** [comment]
As an FYI, I have COSMOS command and telemetry working on the "ground station" VM already, so the help with 42 was the only thing I needed.  I'm interested in seeing your solution since mine may not be the best approach.

From: John P. Lucas ***@***.***>
Sent: Thursday, October 20, 2022 8:05 AM
To: nasa/nos3 ***@***.***>
Cc: Barr, Kathryn A (MSFC-HP25)[MO&I] ***@***.***>; Mention ***@***.***>
Subject: [EXTERNAL] Re: [nasa/nos3] Run NOS3 Example with COSMOS and 42 on different VM (Issue #83)


Final step for this issue is to (from the new dev branch 59e8ec6<https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnasa%2Fnos3%2Fcommit%2F59e8ec6528f0d9220e3776afa90d5026e6e396a8&data=05%7C01%7Ckathryn.a.barr%40nasa.gov%7Cb0b159ce1861438cd7c708dab29ba2c9%7C7005d45845be48ae8140d43da96dd17b%7C0%7C0%7C638018678778900781%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=cbO%2Fo3azFsjzdpE9bsSkIoJAftx9JzL6DwlfkoXR1pE%3D&reserved=0>) recreate the above and get COSMOS command and telemetry working. @msuder<https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fmsuder&data=05%7C01%7Ckathryn.a.barr%40nasa.gov%7Cb0b159ce1861438cd7c708dab29ba2c9%7C7005d45845be48ae8140d43da96dd17b%7C0%7C0%7C638018678778900781%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=j7v6Ja4HcPjYMwDmpvLu29vPGZze8EvFIqiGl44SH1c%3D&reserved=0>

-

**Comment by msuder** [comment]
nos3 repo, branch nos3#83-DualVM (with submodule cosmos branch nos3#83-DualVM) has been configured so that flight software (FSW) and simulators run on one VM with IP address 192.168.100.4 and 42 and COSMOS run on another VM with IP address 192.168.100.5.

The following files/commands have been modified so that FSW/sims run on 192.168.100.4 and 42/COSMOS run on 192.168.100.5:
- sims/cfg/nos_engine_server_config.json
- - configure the nos3 server URI to use 192.168.100.4
- sims/cfg/nos3-simulator.xml
- - configure the nos-connection-string to be 192.168.100.4:12001
- - configure the data-provider hostname to be 192.168.100.5 for each simulator (to connect to 42)
- sims/cfg/InOut/Inp_IPC.txt
- - configure the server hostname to be 192.168.100.5 (for each 42 port that is used for a simulator)
- sims/cfg/InOut/Inp_NOS3.txt
- - configure the NOS3 time connection string to use 192.168.100.4 (42 time connection)
- config/tools/cmd_tlm_server/nos3_cmd_tlm_server.txt
- - configure the MISSION_INT interface to connect to 192.168.100.4 (COSMOS to FSW/CI_LAB/TO_LAB connection)
- COSMOS command:  CFS TO_ENABLE_OUTPUT with DEST_IP '192.168.100.5', DEST_PORT 5011
- - configure FSW/TO_LAB to send telemetry to COSMOS at IP 192.168.100.5

**Comment by msuder** [comment]
Useful for assigning static IPs to VMs:  https://ernieleseberg.com/virtualbox-with-static-ip-on-ubuntu-server-guest/

**Comment by msuder** [comment]
**QUAD VMs:** nos3 repo, branch nos3#83QuadVM (with submodule cosmos branch nos3#83-DualVM) has been configured so that flight software (FSW) and NOS Engine Server run in one VM with IP address 192.168.100.4; sims run in a VM with IP address 192.168.100.5; 42 runs in a VM with IP address 192.168.100.6; and COSMOS runs in a VM with IP address 192.168.100.7.

The following files/commands have been modified for this:
- sims/cfg/nos_engine_server_config.json
-- configure the nos3 server URI to use 192.168.100.4
- sims/cfg/nos3-simulator.xml
-- configure the nos-connection-string to be 192.168.100.4:12001
-- configure the data-provider hostname to be 192.168.100.6 for each simulator (to connect to 42)
-- configure the COSMOS IP for the truth simulator to be 192.168.100.7
- sims/cfg/Inout/Inp_IPC.txt
-- configure the server hosname to be 192.168.100.6 (for each 42 port that is used for a simulator)
- sims/cfg/InOut/Inp_NOS3.txt
-- configure the NOS3 time connection string to use 192.168.100.4:12001 (42 time connection)
- config/tools/cmd_tlm_server/nos3_cmd_tlm_server.txt
- COSMOS command:  CFS TO_ENABLE_OUTPUT with DEST_IP '192.168.100.7', DEST_PORT 5011
-- configure FSW/TO_LAB to send telemetry to COSMOS at IP 192.168.100.7

**Comment by jlucas9** [comment]
Closing for now as new issues will be made to reference this one and further expand on the capability.

**Comment by Simon9050** [comment]
Hi,

First off, big thanks for your work on NOS3 simulator! It’s a game-changer for the space community.

I tried running NOS3 in dual VM mode, following the instructions, but it didn’t quite cooperate... Any tips or a specific branch I should use for this setup?

Cheers, 

Simon
